### ISSUE TITLE
Logger configuration in conf files deprecation message when a logger.resource is supplied

### ISSUE THREAD

**Issue by ricardclau** [playframework/playframework#4574]
Hi

I just upgraded to Play 2.4 from Play 2.3 pretty much flawlessly (BTW brilliant work on the documentation and maintaining compatibility with deprecation warnings!).

In this process, I have removed all the logger references in the conf files to the logback files but for instance, when I run my tests I use the following option in build.sbt

```
javaOptions in Test ++= Seq("-Dconfig.file=conf/application-test.conf", "-Dlogger.resource=test-logger.xml")
```

The tests run fine but I get deprecated messages whenever a fakeApplication is instantiated

I have looked at the code and in *GuiceApplicationBuilder.scala* there is

```
    if (appConfiguration.underlying.hasPath("logger")) {
      Logger.warn("Logger configuration in conf files is deprecated and has no effect. Use a logback configuration file instead.")
    }
```

But of course, this logger path gets defined when I pass the logger.resource option

Is there any other play configuration key to be used or is this just a small bug? I have checked *Logger.scala* and I cannot see how to avoid this issue

Waiting for your comments!

**Comment by marcospereira** [comment]
This is also documented:

1. https://github.com/marcospereira/playframework/blob/master/documentation/manual/detailedTopics/production/ProductionConfiguration.md
2. https://github.com/marcospereira/playframework/blob/master/documentation/manual/detailedTopics/production/ProductionConfiguration.md

Anyway, you could avoid this using different file names for dev environment and production:

https://github.com/marcospereira/playframework/blob/master/framework/src/play/src/main/scala/play/api/Logger.scala#L223-228

In other words, `logback-play-dev.xml` will be used development mode and `logback-play-default.xml` will be used in other modes. Both are fallbacks to `logback.xml` file.

**Comment by ricardclau** [comment]
In my case (and sorry for not mentioning it as well) I have 3 environments running with Play production mode (staging, release and production let's say) so using the same name is not an option for me as I bundle the application and in each environment I pass both the config file and the logback file as an argument. Plus, I run Play Framework as a service. Something like:

```
start on runlevel [2345]
stop on runlevel [!2345]
 
env PLAYFWDIR=/opt/{{ version }}
env PLAYENV={{ env }}
 
script
  $PLAYFWDIR/bin/{{ app }} -Dhttp.port={{ port }} -Dlogger.resource=$PLAYENV-logger.xml -Dconfig.file=$PLAYFWDIR/conf/application-$PLAYENV.conf -J-server -mem {{ jvm_memory }} >> /var/log/playframework/application.log
end script
```

The bug (at least to me) is that the logger key is used to detect if we are using legacy options but the options documented there use a logger key thus causing a deprecation warning.

**Comment by marcospereira** [comment]
@ricardclau I was not suggesting that there is no bug, but just offering an alternative. About the bug, I think we can use `play.logger.(file|resource|url)` as the keys to indicate which logback configuration will be used.

**Comment by ricardclau** [comment]
Oh, no problem at all! I thought I may have not explained properly the situation, sorry about that! 
But yes, unfortunately, the alternative will not work for my case.

I agree that these new keys (prepending the `play` key) would make total sense. Would this be something to be considered for 2.4.1?

**Comment by jroper** [comment]
I think it's simply a bug.  We should be smarter about outputting that deprecation message - check that it looks like configuration for a logger, eg, by checking for values of DEBUG/INFO/WARN/ERROR, and only output the warning if a value of any keys at a subpath of `logger` are equal to that.

As far as `play.logger` vs `logger` is concerned - not sure, note we are talking about something different here, the log file is not read from Play's configuration, it's read from a system property, and therefore doesn't necessarily have the same rules applied for naming.

**Comment by adamhooper** [comment]
This is rather a big deal in functional tests, because they tend to produce log messages that should be ignored.

I've specified a `logback-test.xml` file and I've set it as `-Dlogback.configurationFile=logback-test.xml` ... but starting a `FakeApplication` actually reconfigures logging through Play, so I also need to set `-Dlogger.resource=logback-test.xml`. Because of this issue, every test that uses a FakeApplication spits out this deprecation message.

Eventually, I worked around the problem. I added the `-Dlogger.resource=logback-test.xml` and added this to logback-test.xml:

```xml
  <appender name="STDOUT" class="ch.qos.logback.core.ConsoleAppender">
    <!-- Play 2.4.1 bug: during startup, Play _always_ (even in test) resets
         logger configuration. But it shows a warning if we try to specify
         a log file. https://github.com/playframework/playframework/issues/4574
    -->
    <filter class="ch.qos.logback.core.filter.EvaluatorFilter">
      <evaluator>
        <expression>event.getMessage().contains("Logger configuration in conf files is deprecated and has no effect.")</expression>
      </evaluator>
      <onMatch>DENY</onMatch>
    </filter>
    <encoder>
    ...
```

**Comment by brianwawok** [comment]
I also get this warning despite no logs configured in my .conf files (either dev or prod)...

**Comment by ipsq** [comment]
I'm seeing the same behavior. `application.conf` is almost as clean as it can get and the only play related keys are secret, module and lang, but I still get the `application - Logger configuration in conf files is deprecated and has no effect. Use a logback configuration file instead.` warning message. Funny enough, I'm using a Logback configuration file: `-Dlogger.file=conf/application-logger.xml`.

**Comment by jeffsteinmetz** [comment]
+1 I get the same thing.  application.conf is bare bones.

**Comment by rfranco** [comment]
+1

**Comment by oleg-gr** [comment]
+1 in 2.5.0-M1

**Comment by wsargent** [comment]
So this is here:

https://github.com/playframework/playframework/blob/master/framework/src/play/src/main/scala/play/api/inject/guice/GuiceApplicationBuilder.scala#L93

Fix is:

* Check for logger using hasPath("logger") 
* Iterate through keys that have "logger" in them
* If any of them have values "DEBUG/INFO/WARN/ERROR" then log and bail, otherwise do nothing.

Marking as newbie.

**Comment by dengalepirat** [comment]
If nobody has worked on this, then I have a solution that should work.

Should I write any tests for the solution? Your guidelines doesn't state anything about tests.

**Comment by wsargent** [comment]
@dengalepirat unit tests for the functionality is assumed -- integration tests are nice if you have them.

**Comment by DigitalProf** [comment]
We recently upgraded the Play Framework from 2.2.6 to 2.4.4. We found the jump to be significant, mainly with differences in SQL syntax in Java/Scala/Play. Unfortunately, after the upgrade the logger no longer worked. Since the conversion, we have implemented [the guidance for 2.4.4](https://www.playframework.com/documentation/2.4.4/SettingsLogger), but the log file (logs/application.log) is not being updated. Any ideas about what to look for will be appreciated.

**Comment by tsuyoshizawa** [comment]
We have a PR #5451.
I hope it helps us.

**Comment by wsargent** [comment]
In your logback.xml file, you can also use ${application.home:-.} which will default to the current directory if application.home is not found.  This is a common problem when tests are run directly from an IDE.

**Comment by DigitalProf** [comment]
We will give this a try! Many thanks!

Sent from my iPhone

On Jan 15, 2016, at 12:22 PM, Will Sargent <notifications@github.com<mailto:notifications@github.com>> wrote:


In your logback.xml file, you can also use ${application.home:-.} which will default to the current directory if application.home is not found. This is a common problem when tests are run directly from an IDE.

-

**Comment by wsargent** [comment]
Fixed through https://github.com/playframework/playframework/pull/5451

**Comment by wsargent** [comment]
Also @DigitalProf the application.home default is in https://github.com/playframework/playframework/commit/a8c9f5ea4ef4b32ae3f831f50f2136d86db1530c
